pkg/updater/repository/repository.go

```diff
@@ -0,0 +1,324 @@
+	if err != nil {
+		return err
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	if !repository.stable.Exists() {
+		return fmt.Errorf("stable package does not exist, invalid state")
+	}
+	if !repository.experiment.Exists() {
+		return fmt.Errorf("experiment package does not exist, invalid state")
+	}
+	err = repository.stable.Set(*repository.experiment.packagePath)
+	if err != nil {
+		return fmt.Errorf("could not set stable: %w", err)
+	}
+	err = repository.experiment.Delete()
+	if err != nil {
+		return fmt.Errorf("could not delete experiment link: %w", err)
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	return nil
+}
+
+// DeleteExperiment deletes the experiment.
+//
+// 1. Cleanup the repository.
+// 2. Delete the experiment link.
+// 3. Cleanup the repository to remove the previous experiment package.
+func (r *Repository) DeleteExperiment() error {
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return err
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	if !repository.stable.Exists() {
+		return fmt.Errorf("stable package does not exist, invalid state")
+	}
+	if !repository.experiment.Exists() {
+		return nil
+	}
+	err = repository.experiment.Delete()
+	if err != nil {
+		return fmt.Errorf("could not delete experiment link: %w", err)
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	return nil
+}
+
+type repository struct {
+	rootPath string
+
+	stable     *link
+	experiment *link
+}
+
+func openRepository(rootPath string) (*repository, error) {
+	stableLink, err := newLink(filepath.Join(rootPath, stableVersionLink))
+	if err != nil {
+		return nil, fmt.Errorf("could not load stable link: %w", err)
+	}
+	experimentLink, err := newLink(filepath.Join(rootPath, experimentVersionLink))
+	if err != nil {
+		return nil, fmt.Errorf("could not load experiment link: %w", err)
+	}
+
+	return &repository{
+		rootPath:   rootPath,
+		stable:     stableLink,
+		experiment: experimentLink,
+	}, nil
+}
+
+func (r *repository) setExperiment(name string, sourcePath string) error {
+	path, err := movePackageFromSource(name, r.rootPath, sourcePath)
+	if err != nil {
+		return fmt.Errorf("could not move experiment source: %w", err)
+	}
+	return r.experiment.Set(path)
+}
+
+func (r *repository) setStable(name string, sourcePath string) error {
+	path, err := movePackageFromSource(name, r.rootPath, sourcePath)
+	if err != nil {
+		return fmt.Errorf("could not move stable source: %w", err)
+	}
+	return r.stable.Set(path)
+}
+
+func movePackageFromSource(packageName string, rootPath string, sourcePath string) (string, error) {
+	if packageName == "" || packageName == stableVersionLink || packageName == experimentVersionLink {
```

Comment:
In which cases `packageName` could be `stableVersionLink` or `experimentVersionLink`?

Reply:
If the version of the agent is somehow "stable" or "experiment". This is technically possible so I prefer to handle the edge case.

Reply:
Thanks for the information